src/NodeServices/PythonServices.cs

```diff
@@ -0,0 +1,338 @@
+    [SupressImportIntoVM]
+    public class PythonEngineProxy : IDisposable
+    {
+        /// <summary>
+        /// The version of the Python Engine connected to this proxy instance
+        /// </summary>
+        public readonly PythonEngineVersion Version;
+
+        private readonly Type EngineType;
+        private EvaluationStartedEventHandler EvaluationStartedCallback;
+        private EvaluationFinishedEventHandler EvaluationFinishedCallback;
+
+        /// <summary>
+        /// Add an event handler before the Python evaluation begins
+        /// </summary>
+        /// <param name="callback"></param>
+        public void OnEvaluationBegin(EvaluationStartedEventHandler callback)
+        {
+            EvaluationStartedCallback += callback;
+            AddEventHandler(EvaluationStartedCallback.GetType(), nameof(HandleEvaluationStarted));
+        }
+
+        /// <summary>
+        /// Add an event handler after the Python evaluation ends
+        /// </summary>
+        /// <param name="callback"></param>
+        public void OnEvaluationEnd(EvaluationFinishedEventHandler callback)
+        {
+            EvaluationFinishedCallback += callback;
+            AddEventHandler(EvaluationFinishedCallback.GetType(), nameof(HandleEvaluationFinished));
+        }
+
+        internal PythonEngineProxy(Type eType, PythonEngineVersion version)
+        {
+            EngineType = eType;
+            Version = version;
+        }
+
+        public void Dispose()
+        {
+            try
+            {
+                if (EvaluationStartedCallback != null)
+                {
+                    MethodInfo handlerInfo = typeof(PythonEngineProxy).GetMethod(nameof(HandleEvaluationStarted), BindingFlags.NonPublic | BindingFlags.Instance);
+                    var handler = Delegate.CreateDelegate(typeof(EvaluationStartedEventHandler), this, handlerInfo);
+                    var eventInfo = EngineType.GetEvents().FirstOrDefault(x => x.EventHandlerType == typeof(EvaluationStartedEventHandler));
+                    eventInfo?.RemoveEventHandler(this, handler);
+                }
+
+                if (EvaluationFinishedCallback != null)
+                {
+                    MethodInfo handlerInfo = typeof(PythonEngineProxy).GetMethod(nameof(HandleEvaluationFinished), BindingFlags.NonPublic | BindingFlags.Instance);
+                    var handler = Delegate.CreateDelegate(typeof(EvaluationFinishedEventHandler), this, handlerInfo);
+                    var eventInfo = EngineType.GetEvents().FirstOrDefault(x => x.EventHandlerType == typeof(EvaluationFinishedEventHandler));
+                    eventInfo?.RemoveEventHandler(this, handler);
+                }
+            }
+            catch
+            { }
+        }
+
+        /// <summary>
+        /// Gets the input Marshaler from the target Python Engine
+        /// </summary>
+        /// <returns>DataMarshaler as object</returns>
+        public object GetInputMarshaler()
```

Comment:
I used generic object because if I bring in the datatype DataMarshaler then I either have to add a new dependency on DYnamoUtilities or I have to move the Marshaler class to DynamoServices (and it seems like a good fit for the  Marshaler to be in DynamoUtilities)